src/Utils.vala

```diff
@@ -427,45 +438,189 @@ public class FeedReader.Utils : GLib.Object {
 		return feedname.str;
 	}
 
-	public static bool downloadIconWithSession(Soup.Session session, string feed_id, string feed_url, string icon_path = GLib.Environment.get_user_data_dir() + "/feedreader/data/feed_icons/")
+	public static bool downloadFavIcon(string feed_id, string feed_url, string icon_path = GLib.Environment.get_user_data_dir() + "/feedreader/data/feed_icons/")
 	{
-		var path = GLib.File.new_for_path(icon_path);
-		try{path.make_directory_with_parents();}catch(GLib.Error e){}
-		string local_filename = icon_path + feed_id.replace("/", "_").replace(".", "_") + ".ico";
 
-		string url = feed_url;
-		if(feed_url.has_prefix("http://"))
-			url = url.replace("http://", "");
-		else if(feed_url.has_prefix("https://"))
-			url = url.replace("https://", "");
+		var uri = new Soup.URI(feed_url);
+		string hostname = uri.get_host();
+		int first = hostname.index_of_char('.', 0);
+		int second = hostname.index_of_char('.', first+1);
+		if(second != -1 && first != second)
+			hostname = hostname.substring(first+1);
+		string siteURL = uri.get_scheme() + "://" + hostname;
+
+		// download html and parse to find location of favicon
+		var message_html = new Soup.Message("GET", siteURL);
+		if(Settings.tweaks().get_boolean("do-not-track"))
+			message_html.request_headers.append("DNT", "1");
+		var status = getSession().send_message(message_html);
+		if(status == 200)
+		{
+			var html = (string)message_html.response_body.flatten().data;
+
+			var html_cntx = new Html.ParserCtxt();
+	        html_cntx.use_options(Html.ParserOption.NOERROR + Html.ParserOption.NOWARNING);
+	        Html.Doc* doc = html_cntx.read_doc(html, siteURL, null, Html.ParserOption.NOERROR + Html.ParserOption.NOWARNING);
+	        if(doc == null)
+	        {
+	            Logger.debug("Utils.downloadFavIcon: parsing html failed");
+	    		return false;
+	    	}
+
+			// check for <link rel="apple-touch-icon">
+			var xpath = grabberUtils.getURL(doc, "//link[@rel='apple-touch-icon']");
```

Comment:
So maybe order of precedence should be: `rel="icon"`, `rel="shortcut icon"`, `rel="apple-touch-icon"`. Not that it will make much difference in most cases :D